### DISCUSSION TITLE
Change the colors of an image to that of another one

### DISCUSSION THREAD

**Discussion by rourke** [ImageMagick/ImageMagick#7660] [Help] [answered]
### ImageMagick version

7.1.1-38

### Operating system, version and so on

Arch Linux 6.10.10-arch1-1

### Description

How do I change the colors of a source image in a way that they match another image (in this case a wallpaper)? Unfortunately I can't be more specific because I don't know how to describe this in more details. So it's better to give an example.

**Source image**
![fastfetch](https://github.com/user-attachments/assets/f33cd950-7ae7-4105-83c5-d17c472e6be8)

**A wallpaper**
Source: https://wallhaven.cc/w/j5mz95
![image](https://github.com/user-attachments/assets/cd70fa72-42e5-40ad-93f6-9628ea634470)

**Output image**
This is maybe a bad example, but I tried to visualize my intentions but my photoshop skills are limited. This was mostly a hue/saturation change.
![fastfetch_example](https://github.com/user-attachments/assets/a7084e14-6621-40e4-9a89-98dab5032288)

For what it's worth: I'm doing this through a bash script and I'm able to extract a color palette of the wallpaper through [wallust](https://codeberg.org/explosion-mental/wallust), so I have access to the main colors.

**Comment by fmw42** [comment]
Use my scripts, matchimage or histmatch at http://www.fmwconcepts.com/imagemagick/index.php.  The first tries to adjust the mean and standard deviation from one image to match the other another.  The second tries to adjust one histogram to match the other.

**Reply by rourke** [reply]
Sounds good. But I'm not sure how those scripts are supposed to work.

```shell
matchimage source.png wallpaper.png output1.png
histmatch wallpaper.png output1.png output2.png
```

The first command produces this image:
![output1](https://github.com/user-attachments/assets/a58ed9bd-a3bc-4d31-8720-0f70f4473f30)

The second command produces this image:
![output2](https://github.com/user-attachments/assets/3f0dfc8e-0c0c-41d9-9020-274cc37eec77)

Please note:
- Your scripts ouput this warning, but I suspect this is not an issue?

   ```bash
   WARNING: The convert command is deprecated in IMv7, use "magick" instead of "convert" or "magick convert"
   ```
- When I switch the histmatch arguments 1 and 2 around, output2.png becomes a version of the wallpaper.

**Reply by fmw42** [reply]
Two separate tools that should not necessarily be run in series. Your second command should be as below.

```
matchimage source.png wallpaper.png output1.png
histmatch wallpaper.png source.png output2.png
```

Given that your two images are very different (in colors), you will not be able to match them simply by mean/std matching and histogram matching may be better or worse.

I think you might need an AI approach, which Imagemgick does not do.

**Reply by fmw42** [reply]
Perhaps a better approach would be to get all the unique colors from the destination image and make a colormap image from each. Then use that with -remap to map each of your source colors to the colormap image. Use -dither none. See https://imagemagick.org/Usage/quantize/#remap and https://imagemagick.org/Usage/quantize/#web_safe and https://imagemagick.org/Usage/quantize/#extract

**Reply by fmw42** [reply]
Convert in Imagemagick 7 is deprecated. You might need to make a symbolic link from convert to magick. The message you got is a warning and should not affect the scripts if you are getting results.

**Reply by rourke** [reply, marked as answer]
I tried a few more things, but in the end decided to create a new image with only a few fixed colors, which I then 'remap' when I select a new wallpaper. Basically with this command:

```
magick input.png -fill "${dark}" -opaque "#ff0000" -fill "${lightdark}" -opaque "#00ffff" -fill "${light}" -opaque "#0000ff" output.png
```

Thanks for helping out @fmw42.
